src/treetensornetwork/abstractprojttno.jl

```diff
@@ -0,0 +1,145 @@
+abstract type AbstractProjTTNO{V} end
+
+copy(::AbstractProjTTNO) = error("Not implemented")
+
+set_nsite(::AbstractProjTTNO, nsite) = error("Not implemented")
+
+# silly constructor wrapper
+shift_position(::AbstractProjTTNO, pos) = error("Not implemented")
+
+make_environment!(::AbstractProjTTNO, psi, e) = error("Not implemented")
+
+underlying_graph(P::AbstractProjTTNO) = underlying_graph(P.H)
+
+pos(P::AbstractProjTTNO) = P.pos
+
+Graphs.edgetype(P::AbstractProjTTNO) = edgetype(underlying_graph(P))
+
+on_edge(P::AbstractProjTTNO) = isa(pos(P), edgetype(P))
+
+nsite(P::AbstractProjTTNO) = on_edge(P) ? 0 : length(pos(P))
+
+function sites(P::AbstractProjTTNO{V}) where {V}
+  on_edge(P) && return V[]
+  return pos(P)
+end
+
+function incident_edges(P::AbstractProjTTNO{V})::Vector{NamedEdge{V}} where {V}
+  on_edge(P) && return [pos(P), reverse(pos(P))]
+  edges = [
+    [edgetype(P)(n => v) for n in setdiff(neighbors(underlying_graph(P), v), sites(P))] for
+    v in sites(P)
+  ]
+  return collect(Base.Iterators.flatten(edges))
+end
+
+function internal_edges(P::AbstractProjTTNO{V})::Vector{NamedEdge{V}} where {V}
```

Comment:
This is a good idea for a generic graph function. It is the complement of the edge boundary of a subgraph (https://en.wikipedia.org/wiki/Boundary_(graph_theory) and `boundary_edges` added in https://github.com/mtfishman/NamedGraphs.jl/pull/20).

As we discussed, a simple implementation would be `internal_edges(g::AbstractGraph, vertices) = edges(subgraph(g, vertices))`. Then this function could be defined as `internal_edges(P::AbstractProjTTNO) = internal_edges(underlying_graph(P), sites(P))`.